### PR TITLE
🐇 Avoid passing options to formatting rules

### PR DESCRIPTION
**Pull request by sharwell**
With this change, the rule instance is responsible for the keeping track of the correct options. This avoids the need to repeatedly look up options during formatting, and reduces the invocation costs associated with chained operations.

### CODE REVIEW

File: src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs [hunk 1]
```diff
@@ -193,27 +193,45 @@ private int GetDesiredIndentation(IBraceCompletionSession session, ITextSnapshot
...
                 Debug.Assert(s_instances[(int)indentStyle]._indentStyle == indentStyle);
                 return s_instances[(int)indentStyle];
             }
 
-            public override AdjustNewLinesOperation GetAdjustNewLinesOperation(SyntaxToken previousToken, SyntaxToken currentToken, AnalyzerConfigOptions options, in NextGetAdjustNewLinesOperation nextOperation)
+            public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options)
+            {
+                var newLinesForBracesInObjectCollectionArrayInitializers = options.GetOption(CSharpFormattingOptions.NewLinesForBracesInObjectCollectionArrayInitializers);
+
+                if (newLinesForBracesInObjectCollectionArrayInitializers == _newLinesForBracesInObjectCollectionArrayInitializers)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs:224]: 1. cute :)   maybe add explanatory comment on the field?
2. how many values can this option have? It's a bool right?  for people not on the default, can we cache that as well?  so just have two cached rules and only return one or the other.

Comment (sharwell) [reply src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs:224]: This was the old pattern. I meant to convert them all to structs.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs [hunk 2]
```diff
@@ -11,28 +11,53 @@
...
+        private IndentBlockFormattingRule(CachedOptions options)
+        {
+            _options = options;
+        }
+
+        public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options)
+        {
+            var cachedOptions = new CachedOptions(options);
+
+            if (cachedOptions == _options)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:42]: just to check, but we're overriding == then?

Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:42]: Yes, it's a struct with an overloaded operator.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs [hunk 3]
```diff
@@ -11,28 +11,53 @@
...
+            {
+                return this;
+            }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:45]: so we basically reuse if user hasn't change from the defaults.  but we always allocate if they have?  woudl def like an approach where we can get the benefits as well for people not on the defaults.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs [hunk 4]
```diff
@@ -299,5 +320,60 @@ private void AddEmbeddedStatementsIndentationOperation(List<IndentBlockOperation
...
+
+            public override int GetHashCode()
+            {
+                var hashCode = 0;
+                hashCode = (hashCode << 2) + (int)LabelPositioning;
+                hashCode = (hashCode << 1) + (IndentBlock ? 1 : 0);
+                hashCode = (hashCode << 1) + (IndentSwitchCaseSection ? 1 : 0);
+                hashCode = (hashCode << 1) + (IndentSwitchCaseSectionWhenBlock ? 1 : 0);
+                hashCode = (hashCode << 1) + (IndentSwitchSection ? 1 : 0);
+                return hashCode;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:375]: no HashCode.Combine?

Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:375]: This is faster. Even if a conflict happens the hash will be fast.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs (hunk 3 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:45]: In my measurements it didn't matter. <0.1%.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:45]: if that's the case, why do we even need the "return this" optimization?

Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:45]: It's the standard pattern for 'with' methods.

Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/IndentBlockFormattingRule.cs:45]: Fair enough. :)

File: src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/CurlyBraceCompletionSession.cs]: Fixed now

**Comment by CyrusNajmabadi** [comment]
> With this change, the rule instance is responsible for the keeping track of the correct options. This avoids the need to repeatedly look up options during formatting, and reduces the invocation costs associated with chained operations.

I'm curious overall what sort of savings we see.  Thanks!

**Comment by sharwell** [comment]
Here's the command I used for testing:

> .\artifacts\bin\AnalyzerRunner\Release\netcoreapp3.1\AnalyzerRunner.exe .\artifacts\bin\Microsoft.CodeAnalysis.CSharp.CodeStyle\Release\netstandard2.0\ .\Roslyn.sln /stats /concurrent /a CSharpFormattingAnalyzer /iter:3

Here's two executions before this change:

```
Found 0 diagnostics in 77092ms (80632639536 bytes allocated)
Execution time (ms):            1309474.5278

Found 0 diagnostics in 76116ms (80820867824 bytes allocated)
Execution time (ms):            1516483.838
```

Here's two executions after this change:

```
Found 0 diagnostics in 64012ms (79678407728 bytes allocated)
Execution time (ms):            1061650.1599

Found 0 diagnostics in 69589ms (80361379872 bytes allocated)
Execution time (ms):            1255031.3079
```

**Comment by CyrusNajmabadi** [comment]
That's a great improvement! :)

**Pull request merged**